File: algorithms/cpp/validNumber/validNumber.cpp
// Source: https://leetcode.com/problems/valid-number/
// Author: Miao Zhang
// Date:   2021-01-13

class Solution {
public:
    bool isNumber(string s) {
        if (s.empty()) {
            return false;
        }
        int i = 0, j = s.size() - 1;
        while (i <= j && s[i] == ' ') i++;
        while (j >= 0 && s[j] == ' ') j--;
        bool is_digit = false, is_dot = false, is_e = false;
        for (;i <= j; i++) {
            if (s[i] >= '0' && s[i] <= '9') {
                is_digit = true;
            } else if (s[i] == '-' || s[i] == '+') {
                if (i != 0 && s[i - 1] != 'e' && s[i - 1] != 'E') {
                    return false;
                }
            } else if (s[i] == '.') {
                if (is_dot || is_e) {
                    return false;
                }
                is_dot = true;
            } else if (s[i] == 'e' || s[i] == 'E') {
                if (not is_digit || is_e) {
                    return false;
                }
                is_digit = false;
                is_e = true;
            } else {
                return false;
            }
        }
        return is_digit;
    }
};

File: algorithms/python/validNumber/validNumber.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/valid-number/
# Author: Miao Zhang
# Date:   2021-01-13

class Solution:
    def isNumber(self, s: str) -> bool:
        s = s.strip()
        is_digit, is_dot, is_e = False, False, False
        
        for i in range(len(s)):
            if s[i].isdecimal():
                is_digit = True
            elif s[i] in '+-':
                if i != 0 and s[i - 1] != 'e':
                    return False
            elif s[i] == '.':
                if is_dot or is_e:
                    return False
                is_dot = True
            elif s[i].lower() == 'e':
                if not is_digit or is_e:
                    return False
                is_digit = False
                is_e = True
            else:
                return False
        return is_digit
